src/rust/engine/logging/src/logger.rs

```diff
@@ -171,107 +146,62 @@ impl Log for PantsLogger {
-        );
-
-        let level = record.level();
-        let use_color = self.use_color.load(Ordering::SeqCst);
-
-        let level_marker = match level {
-          _ if !use_color => format!("[{}]", level).normal().clear(),
-          Level::Info => format!("[{}]", level).normal(),
-          Level::Error => format!("[{}]", level).red(),
-          Level::Warn => format!("[{}]", level).red(),
-          Level::Debug => format!("[{}]", level).green(),
-          Level::Trace => format!("[{}]", level).magenta(),
-        };
+    let cur_date = chrono::Local::now();
+    let time_str = format!(
+      "{}.{:02}",
+      cur_date.format(TIME_FORMAT_STR),
+      cur_date.time().nanosecond() / 10_000_000 // two decimal places of precision
+    );
+
+    let level = record.level();
+    let destination_is_file = match destination {
+      Destination::Pantsd => true,
+      Destination::Stderr => false,
+    };
+    let use_color = self.use_color.load(Ordering::SeqCst) && (!destination_is_file);
+
+    let level_marker = match level {
+      _ if !use_color => format!("[{}]", level).normal().clear(),
+      Level::Info => format!("[{}]", level).normal(),
+      Level::Error => format!("[{}]", level).red(),
+      Level::Warn => format!("[{}]", level).red(),
+      Level::Debug => format!("[{}]", level).green(),
+      Level::Trace => format!("[{}]", level).magenta(),
+    };
 
-        let log_string = format!("{} {} {}", time_str, level_marker, record.args());
+    let log_string = format!("{} {} {}", time_str, level_marker, record.args());
 
-        {
-          // We first try to output to all registered handlers. If there are none, or any of them
-          // fail, then we fallback to sending directly to stderr.
-          let handlers_map = self.stderr_handlers.lock();
-          let mut any_handler_failed = false;
-          for callback in handlers_map.values() {
-            let handler_res = callback(&log_string);
-            if handler_res.is_err() {
-              any_handler_failed = true;
-            }
+    match destination {
+      Destination::Stderr => {
+        // We first try to output to all registered handlers. If there are none, or any of them
+        // fail, then we fallback to sending directly to stderr.
+        let handlers_map = self.stderr_handlers.lock();
+        let mut any_handler_failed = false;
+        for callback in handlers_map.values() {
+          let handler_res = callback(&log_string);
+          if handler_res.is_err() {
+            any_handler_failed = true;
           }
-          if handlers_map.len() == 0 || any_handler_failed {
-            self.stderr_log.lock().log(record);
+        }
+        if handlers_map.len() == 0 || any_handler_failed {
+          eprintln!("{}", log_string);
+        }
+      }
+      Destination::Pantsd => {
+        let mut maybe_file = self.log_file.lock();
+        if let Some(ref mut file) = *maybe_file {
+          match writeln!(file, "{}", log_string) {
```

Comment:
Why not just use [`write_all`](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all)? Won't the macro copy `log_string` and then write (which is redundant)?

Reply:
I don't think so. `writeln!` invokes `write!` which in turn invokes the `format_args` macro (https://doc.rust-lang.org/src/core/macros/mod.rs.html#414-416), which is a compiler builtin (https://doc.rust-lang.org/src/core/macros/mod.rs.html#789-792). Then `Write::write_fmt` is called with the `Arguments` structure created by the format_args(  https://doc.rust-lang.org/src/core/macros/mod.rs.html#789-792). I don't think any of the types involved in this process are actually cloning `log_string`, and even if they were it's something I think the compiler is smart enough to optimize out.
`write_all` would also work for writing the string to the file, this way just makes it marginally easier to include the trailing newline.

Reply:
It seems unnecessary though to invoke the macro and its attendant logic when you could just `write_all`. Regarding the newline, just adding a `\n` to the `let log_string = format!("{} {} {}", time_str, level_marker, record.args());` would take care of that, correct?

Reply:
(versus reasoning about what the macro may or may not do)

Reply:
> It seems unnecessary though to invoke the macro and its attendant logic when you could just `write_all`. Regarding the newline, just adding a `\n` to the `let log_string = format!("{} {} {}", time_str, level_marker, record.args());` would take care of that, correct?

Yeah but that would add a newline in the dynamic UI case, which we don't want. It would be easy enough to just have two calls to `write_all` in the `Destination::Pantsd` block of the match, where the second one is just `write_all(&['\n' as u8])`. But this is exactly the sort of thing the `writeln!` macro is meant for, and in this case I think it makes the code clearer.

Reply:
Ack.